src/ui/edit_field.c

```diff
@@ -20,14 +20,270 @@ struct Edit_field
     Color font_color;
 };
 
-static void edit_field_left(Edit_field *edit_field);
-static void edit_field_right(Edit_field *edit_field);
-static void edit_field_home(Edit_field *edit_field);
-static void edit_field_end(Edit_field *edit_field);
-static void edit_field_backspace(Edit_field *edit_field);
-static void edit_field_delete(Edit_field *edit_field);
 static void edit_field_insert_char(Edit_field *edit_field, char c);
 
+// See: https://www.gnu.org/software/emacs/manual/html_node/emacs/Moving-Point.html
+// For an explanation of the naming terminology for these helper methods
+static bool is_emacs_word(char c);
+static void forward_char(Edit_field *edit_field);
+static void backward_char(Edit_field *edit_field);
+static void move_beginning_of_line(Edit_field *edit_field);
+static void move_end_of_line(Edit_field *edit_field);
+static void forward_word(Edit_field *edit_field);
+static void backward_word(Edit_field *edit_field);
+static void delete_char(Edit_field *edit_field);
+static void delete_backward_char(Edit_field *edit_field);
+static void kill_word(Edit_field *edit_field);
+static void backward_kill_word(Edit_field *edit_field);
+
+static void handle_keydown(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_alt(Edit_field *edit_field, const SDL_Event *event);
+static void handle_keydown_ctrl(Edit_field *edit_field, const SDL_Event *event);
+
+static void edit_field_insert_char(Edit_field *edit_field, char c)
+{
+    if (edit_field->buffer_size >= BUFFER_CAPACITY) {
+        return;
+    }
+
+    for (int64_t i = (int64_t) edit_field->buffer_size - 1; i >= (int64_t) edit_field->cursor; --i) {
+        edit_field->buffer[i + 1] = edit_field->buffer[i];
+    }
+
+    edit_field->buffer[edit_field->cursor++] = c;
+    edit_field->buffer[++edit_field->buffer_size] = 0;
+}
+
+// See: https://www.gnu.org/software/emacs/manual/html_node/emacs/Moving-Point.html
+// For an explanation of the naming terminology for these helper methods
+
+static bool is_emacs_word(char c)
+{
+#define IN_RANGE(start, end) (c >= start && c <= end)
+
+    // Word syntax table retrieved from Fundamental Mode, "C-h s"
+    // (This is not the complete syntax table)
+    return IN_RANGE('$', '%')
+        || IN_RANGE('0', '9')
+        || IN_RANGE('A', 'Z')
+        || IN_RANGE('a', 'z')
+        ;
```

Comment:
Thank you for making my eyes to do an extra jump at the top of the function. Now, please, inline the macro and remove it completely, so nobody has to do that again.